File: desktop/src/utils/valuation.test.mjs
import assert from 'node:assert';
import fs from 'node:fs';
import { CONDITIONS, EDITIONS, EDITION_LABELS, conditionFactor, unitPrice, valueOf, firstEdLine, groupCopies } from './valuation.js';

// Spec G4 §6/§7 — ZWILLING: desktop/electron/valuation.test.cjs und android ValuationTest.kt lesen dieselbe Fixture.
const FIX = JSON.parse(fs.readFileSync(new URL('../../../docs/fixtures/valuation/first-ed.json', import.meta.url), 'utf8'));

assert.deepStrictEqual(CONDITIONS, ['MT', 'NM', 'EX', 'GD', 'LP', 'PL', 'PO']);
assert.deepStrictEqual(EDITIONS, ['first', 'unlimited', 'limited', 'unknown']);
assert.equal(EDITION_LABELS.first, '1st Ed');
assert.equal(EDITION_LABELS.unknown, 'Unbek.');
assert.equal(conditionFactor('LP'), 0.5);
assert.equal(valueOf({ price: 4 }, [{ condition: 'NM' }, { condition: 'EX' }]), 7.4);

for (const c of FIX.unitPrice) assert.strictEqual(unitPrice(c.card, c.copy), c.unit, c.name);
for (const c of FIX.valueOf) assert.strictEqual(valueOf(c.card, c.copies), c.value, c.name);
// Intl setzt ein geschuetztes Leerzeichen vor das Euro-Zeichen; die Fixture schreibt ein normales.
for (const c of FIX.line) {
  const got = firstEdLine(c.card);
  assert.strictEqual(got == null ? null : got.replace(/ /g, ' '), c.line, c.name);
}

const groups = groupCopies([
  { copy_id: 'a', edition: 'unknown', condition: 'NM' },
  { copy_id: 'b', edition: 'first', condition: 'GD' },
  { copy_id: 'c', edition: 'unknown', condition: 'NM' },
]);
assert.deepStrictEqual(groups, [
  { edition: 'first', condition: 'GD', count: 1 },
  { edition: 'unknown', condition: 'NM', count: 2 },
]);
console.log('valuation renderer test: PASS');

File: android/app/src/test/java/com/example/yugiohscanner/ValuationTest.kt
package com.example.yugiohscanner

import com.example.yugiohscanner.cloud.CardRow
import com.example.yugiohscanner.cloud.CopyRow
import com.example.yugiohscanner.cloud.Valuation
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/** Spec G4 §6 -- ZWILLING von desktop/electron/valuation.test.cjs und desktop/src/utils/valuation.test.mjs (Fixture docs/fixtures/valuation/first-ed.json). */
class ValuationTest {
    private val fix = JSONObject(Fixtures.text("docs/fixtures/valuation/first-ed.json"))

    private fun copy(cond: String, ed: String = "unknown", id: String = "id-$cond-$ed-${System.nanoTime()}") =
        CopyRow(id, "1", "LOB-DE001", "DE", "Ultra Rare", ed, cond, false,
            containerId = null, page = null, slot = null, tags = null, note = null)

    private fun dbl(o: JSONObject, k: String): Double? = if (o.isNull(k)) null else o.getDouble(k)

    private fun card(o: JSONObject) = CardRow("1", "LOB-DE001", "DE", "Test", null, "Ultra Rare", 1, dbl(o, "price"),
        priceFirstEd = dbl(o, "price_first_ed"), cmFirstEdFactor = dbl(o, "cm_first_ed_factor"))

    private fun plain(price: Double?) = CardRow("1", "LOB-DE001", "DE", "Test", null, "Ultra Rare", 1, price)

    @Test
    fun factorsMatchTheSharedJsonFile() {
        // Gradle runs unit tests with the module dir (android/app) as working directory.
        val f = File("../../desktop/electron/condition-factors.json")
        assertTrue("shared factor file must exist at ${f.absolutePath}", f.exists())
        val json = JSONObject(f.readText())
        for (c in Valuation.CONDITIONS) assertEquals("factor $c", json.getDouble(c), Valuation.factor(c), 1e-9)
        assertEquals(json.length(), Valuation.CONDITIONS.size)
    }

    @Test
    fun valueOfSumsUnitPriceTimesFactor() {
        assertEquals(17.0, Valuation.valueOf(plain(10.0), listOf(copy("NM"), copy("GD"))), 1e-9)
        assertEquals(0.0, Valuation.valueOf(plain(null), listOf(copy("NM"))), 1e-9)
        assertEquals(1.0, Valuation.factor(""), 1e-9)
        assertEquals(1.0, Valuation.factor("XX"), 1e-9)
        assertEquals("geloeschte Exemplare zaehlen nicht", 10.0,
            Valuation.valueOf(plain(10.0), listOf(copy("NM"), copy("NM").copy(deleted = true))), 1e-9)
    }

    @Test
    fun `Fixture unitPrice`() {
        val cases = fix.getJSONArray("unitPrice")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val cp = c.getJSONObject("copy")
            assertEquals(c.getString("name"), c.getDouble("unit"),
                Valuation.unitPrice(card(c.getJSONObject("card")), copy(cp.getString("condition"), cp.getString("edition"))), 1e-9)
        }
    }

    @Test
    fun `Fixture valueOf`() {
        val cases = fix.getJSONArray("valueOf")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val arr = c.getJSONArray("copies")
            val copies = (0 until arr.length()).flatMap { j ->
                val o = arr.getJSONObject(j)
                (0 until o.optInt("count", 1)).map { k -> copy(o.getString("condition"), o.getString("edition"), "c$j-$k") }
            }
            assertEquals(c.getString("name"), c.getDouble("value"), Valuation.valueOf(card(c.getJSONObject("card")), copies), 1e-9)
        }
    }

    @Test
    fun `Fixture Preiszeile`() {
        val cases = fix.getJSONArray("line")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val expected = if (c.isNull("line")) null else c.getString("line")
            assertEquals(c.getString("name"), expected, Valuation.firstEdLine(card(c.getJSONObject("card"))))
        }
    }

    @Test
    fun groupOrdersByEditionThenCondition() {
        val g = Valuation.group(listOf(copy("NM"), copy("GD", "first"), copy("NM")))
        assertEquals(listOf(Valuation.Group("first", "GD", 1), Valuation.Group("unknown", "NM", 2)), g)
    }
}
